### ISSUE TITLE
More multi-arch architecures: arm32, riscv etc

### ISSUE THREAD

**Issue by wader** [wader/static-ffmpeg#217]
Current issues are:

- Problem finding hosting provides providing native ARM32, RISCV etc hosts
- Cross compile is messy
- Emulated docker builds a bit less messy but have issues:
  - Emulation is user-land only so `/proc/cpuinfo` etc from the host leaks into the container confusing some build systems (ex: aom ARM32 build gets confused about CPU features, I think it looks as cpuinfo)
  - How to verify that successful build actually was built with CPU features we want?
- Emulated build is slow and uses lots of resources
  - Current static-ffmpeg with all features can't be build on github action host. Runs out of memory and would probably also run out of CPU time (6h max jobs time). Biggest memory/CPU user is rust it seems, without that it might work.

Some details and attempts here https://github.com/wader/static-ffmpeg/pull/204

**Comment by wader** [comment]
@mathieu-aubin let's move arch discussion here

> did you read this -> https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/
> 
> and of so, is it relevant? not much time npw but i figured id input

Yeap docker buildx is one thing i'vet tried and it works similar to how the multi-arch images work, does emulation in user-land. It has the issues i described in above.

> this one too -> https://hub.docker.com/r/linuxserver/ffmpeg

Had a quick look, would guess they build on native hosts or on fully emulated hosts for each arch https://github.com/linuxserver/docker-ffmpeg/blob/226ddf020af76b99b71dbe3804cfca1919f83c11/Jenkinsfile#L466

It would be interesting to do a github action that boots up a full emulated host per arch. I did some experiments with this once and building worked fine for arm32 etc. The messy part is probably to make it nice to work from a workflow, maybe ideally it could be an action that kind of is a docker action that transfers the job env and runs docker inside the guest.

But sadly if run on a github action host it will still run into the same resources limits that i mentioned above.

**Comment by gtsteffaniak** [comment]
I maintain https://github.com/gtsteffaniak/filebrowser which I recently started using this image as a source.

I tappreciate the statically linked binaries and recent ffmpeg version support this repo offers, which is why I chose it. However, it is problematic that it doesn't have armv7 support.

I have lots of multi arch docker experience and it's not hard at all with buildx. There's a lot that could actually be simplified by using it.

I would love to contribute in order to add support for arm32. Since that's missing I can't fully use this image for all my docker images.

Let me know if you would be open to a contribution in that way.

**Comment by wader** [comment]
> I maintain https://github.com/gtsteffaniak/filebrowser which I recently started using this image as a source.

🥳 

> I appreciate the statically linked binaries and recent ffmpeg version support this repo offers, which is why I chose it. However, it is problematic that it doesn't have armv7 support.

Yeap i would like to support it too if we could figure out a good way of doing it.

> I have lots of multi arch docker experience and it's not hard at all with buildx. There's a lot that could actually be simplified by using it.

As i mentioned above i the issue description i have done some experiments using emulated docker buildx but ran into som issues, slowness, hitting resource limits and confusing about target arch, that made me not very confident in the approach. But maybe things have change? at least the slowness i guess could be fixed now by using a arm64 host to run/emulate arm32?

> I would love to contribute in order to add support for arm32. Since that's missing I can't fully use this image for all my docker images.
> 
> Let me know if you would be open to a contribution in that way.

That would be great!

I think the solution with the least amount changes, maintenance and future headaches would be to somehow build on a action runner that is native arm32. At the moment i use ubicloud for arm64 but i don't see arm32 in their documentation but maybe there is ways to do it?

**Comment by gtsteffaniak** [comment]
Thanks for the reply - It is a very reliable and consistent process, but you might be on to something about some timeouts for the emulated versions, because they do take 3-4 times longer and I see your builds are 1h+ right now.

But, the way you are doing multiarch could be improved in both performance and simplicity. I will experiment and make a pull request later.

**Comment by wader** [comment]
> Thanks for the reply - It is a very reliable and consistent process, but you might be on to something about some timeouts for the emulated versions, because they do take 3-4 times longer and I see your builds are 1h+ right now.

Yeap if i remember correctly emulated rust builds are very slow and i think also sometimes runs out of memory on github action runners. But what i'm mostly concern about is that docker buildx is a user-land only emulation so some details of the actual host will leak thru via `/proc` etc which in turn can confuse compilers and build systems unless you explicitly tell them what to target. If done that way i think it would be good to somehow verify the binaries that they really are only the target ISA so that later on when updating some library incompatible instructions sneaks in, this has even happen when doing totally native builds (ex https://github.com/fraunhoferhhi/vvenc/issues/493)

> But, the way you are doing multiarch could be improved in both performance and simplicity. I will experiment and make a pull request later.

Ok! one think that i do value with the current Dockerfile is that it is straight forward and quite easy to follow so it would be great to not introduce too exotic stuff.

**Comment by gtsteffaniak** [comment]
Yeah I really like how simple it is, I'm not looking to rock the boat with any major changes, just get it building with arm32 as well.

You might be right that the user space emulation could cause some issues. I'll see, qemu has improved a lot, but mostly for arm64 due to apple m series.

One question I had, why do you download and build a lot of the build dependencies from source statically? My thinking was ffmpeg was the only thing that needs to be built statically. And besides version controlling, downloading and building everything before ffmpeg is 90% of the build time.

**Comment by wader** [comment]
> Yeah I really like how simple it is, I'm not looking to rock the boat with any major changes, just get it building with arm32 as well.

ideally for me would be more or less no changes and somehow get a native arm32 action runner env somehow.

> You might be right that the user space emulation could cause some issues. I'll see, qemu has improved a lot, but mostly for arm64 due to apple m series.

Yeap i would imagine that arm32 emulated on arm64 could be very fast? but i haven't tried.

> One question I had, why do you download and build a lot of the build dependencies from source statically? My thinking was ffmpeg was the only thing that needs to be built statically. And besides version controlling, downloading and building everything before ffmpeg is 90% of the build time.

You mean why not use alpine static packages? or you mean download binary versions of the library somehow? there is a mix of reasons. Why no alpine: because sometimes the alpine package is not updated as fast, sometimes there is no static package (some i've filed bugs about), sometimes i want to build the library a bit differently. Why from source: Some libraries i want to build with hardened flags (especially decoders), sometimes there is no alpine package. So it's a bit of a mix bag, no general rule really.

**Comment by gtsteffaniak** [comment]
Yeah I am understanding it a bit more as I attempt to build .

Looks like there's a few libraries that have issues , I still need to see if I can figure out support:

- `cairo` build always fails only on armv7, needed for `librsvg`, so for now I have that disabled.
- `aom` requires disabling CPU optimizations for arm32, so I added a condition fro that:

```
# build with generic CPU for armv7l. see croos compiling https://aomedia.googlesource.com/aom
RUN cd aom && \
  if [[ $(uname -m) == "armv7l" ]]; then GENERIC_CPU="-DAOM_TARGET_CPU=generic " else GENERIC_CPU="" fi && \
  mkdir build_tmp && cd build_tmp && \
  cmake \
    -G"Unix Makefiles" \
    -DCMAKE_VERBOSE_MAKEFILE=ON \
    -DCMAKE_BUILD_TYPE=Release \
    -DBUILD_SHARED_LIBS=OFF \
    -DENABLE_EXAMPLES=NO \
    -DENABLE_DOCS=NO \
    -DENABLE_TESTS=NO \
    -DENABLE_TOOLS=NO \
    -DCONFIG_TUNE_VMAF=1 \
    -DENABLE_NASM=ON \
    -DCMAKE_INSTALL_LIBDIR=lib \
    $GENERIC_CPU \
  .. && \
  make -j$(nproc) install
```

- There's also an issue with `xavs2` on arm32, still looking at whether that can be fixed

**Comment by gtsteffaniak** [comment]
I am testing on an m4 MacBook , while the compile times are faster, arm32 emulation is quite slow in general -- but it does work with qemu. I will test the ci builds to see how long they take, locally armv7 builds take ~20 minutes.

Also, since buildx all happens on the same node, I am using a prefetch sources script to download all the versions once beforehand and unzip them. This should also save some time since it happens on the native node.

Once the branch I have is ready with the optimizations , ill let you decide how you want to include it.

But I'll raise a PR to do a simple update to the dockerfile to support arm32 and update the GitHub action to use buildx.

**Comment by wader** [comment]
> Yeah I am understanding it a bit more as I attempt to build .
> 
> Looks like there's a few libraries that have issues , I still need to see if I can figure out support:
> 
> * `cairo` build always fails only on armv7, needed for `librsvg`, so for now I have that disabled (only for armv7)

Any idea what it fails on? the librsvg build system is quite complex unfortunately, meson with python scripts that wrap cargo 😬 maybe arm64 build host feature leaking into target?

> * `aom` requires disabling CPU optimizations for arm32, so I added a condition for that:

Maybe same as above about arm64 build host features leaks into target?

> ```
> # build with generic CPU for armv7l. see cross compiling https://aomedia.googlesource.com/aom
> RUN cd aom && \
>   if [[ $(uname -m) == "armv7l" ]]; then GENERIC_CPU="-DAOM_TARGET_CPU=generic " else GENERIC_CPU="" fi && \
>   mkdir build_tmp && cd build_tmp && \
>   cmake \
>     -G"Unix Makefiles" \
>     -DCMAKE_VERBOSE_MAKEFILE=ON \
>     -DCMAKE_BUILD_TYPE=Release \
>     -DBUILD_SHARED_LIBS=OFF \
>     -DENABLE_EXAMPLES=NO \
>     -DENABLE_DOCS=NO \
>     -DENABLE_TESTS=NO \
>     -DENABLE_TOOLS=NO \
>     -DCONFIG_TUNE_VMAF=1 \
>     -DENABLE_NASM=ON \
>     -DCMAKE_INSTALL_LIBDIR=lib \
>     $GENERIC_CPU \
>   .. && \
>   make -j$(nproc) install
> ```
> 
> * There's also an issue with `xavs2` on arm32, still looking at whether that can be fixed

Okok, looks like i already disable some asm stuff for xavs2, is probably assume amd64 etc so probably ok with more hacks.

As i mention one thing that would be great would be to somehow test or verify that binaries that they actually are using only the ISA they are suppose to use, but not sure how that could be done. I might take a look into that, should maybe be done separate from this?

**Comment by gtsteffaniak** [comment]
I noticed the post check was failing for armv7l :

```
RUN \
  /checkelf /usr/local/bin/ffmpeg && \
  /checkelf /usr/local/bin/ffprobe

0.210 /usr/local/bin/ffmpeg: ELF 32-bit LSB pie executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-armhf.so.1, BuildID[sha1]=a99c4a15b2def3452d3ba47ff95c37b9195b6c8d, stripped
```

This is specific to ffmpeg compilation, even when just compiling ffmpeg without enabling anything its an issue:

```
ARG CFLAGS="-static-libgcc -fno-strict-overflow -fstack-protector-all -fPIC"
ARG CXXFLAGS="-static-libgcc -fno-strict-overflow -fstack-protector-all -fPIC"
RUN cd ffmpeg* && \
  ./configure \
  --pkg-config-flags="--static" \
  --extra-cflags="$CFLAGS" \
  --extra-cxxflags="$CXXFLAGS" \
  --extra-ldexeflags="-fPIE -static-pie" \
  --extra-ldflags="-fopenmp -Wl,--allow-multiple-definition -Wl,-z,stack-size=2097152" \
  --disable-shared \
  --disable-ffplay \
  --enable-static \
  || (cat ffbuild/config.log ; false) && \
  make -j$(nproc) install
```

The above always results in a dynamically linked binary against `/lib/ld-musl-armhf.so.1` and I don't see anything I am doing wrong.

Do you have any ideas? (I removed the sed and toolchain=hardened with the actual flags). It works with amd64 and arm64, just not armv7l

**Comment by wader** [comment]
First thanks for digging into this! yeap the wonderful world of static linking, different targets and weird build systems 😬 Cross compile could be an option, what i fear about is the mess of trying to convince all the different build systems to behave and to not mess up important hardware optimisations etc, but it might be worth a try.

Hope i can get some time tomorrow to look into it the `ld-musl-armhf.so.1` thingy, it might be worth search around a bit if other have had similar issues with static musl on arm7l? what's the difference between armv7 and armv7l btw? i tried to look it up but just got even more confused

**Comment by gtsteffaniak** [comment]
I think there is a difference, but armv7l is just the name the kernel uses for armv7, I think the technical difference is the name of the instructions set vs the name of the architecture.

Armv8 is aarch64 ... And armv8.1 is also aarch64

Yeah it's a good question about optimizations. Would be worth doing maybe just for armv7 since right now there's nothing

**Comment by gtsteffaniak** [comment]
That looks like better output, but does that mean the checkelf fails?

One easy test is to add another FROM ubuntu block and copy the binaries and try to run them in the next step

**Comment by wader** [comment]
Good point, will check tomorrow. If i remember correctly static-pie binaries do for some reason output the "builtin" linker in the ldd output.

Think i've found a shorter reproduction:

armv7
```sh
$ docker run -ti --rm --platform linux/arm/v7 alpine:3.20.3 sh -c 'apk add build-base && gcc -x c -Wl,--verbose=2 -v -static-pie <(echo "int main() {return 0;}") -o test && ldd test && file test'
...
attempt to open /usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../libc.so succeeded
...
/usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../libc.so
attempt to open /usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/crtendS.o succeeded
/usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/crtendS.o
attempt to open /usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../crtn.o succeeded
/usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../crtn.o
COLLECT_GCC_OPTIONS='-v' '-static-pie' '-o' 'test'  '-mtune=generic-armv7-a' '-mfloat-abi=hard' '-mfpu=vfpv3-d16' '-mabi=aapcs-linux' '-mtls-dialect=gnu' '-mthumb' '-mlibarch=armv7-a+fp' '-march=armv7-a+fp' '-dumpdir' 'test.'
	/lib/ld-musl-armhf.so.1 (0x400000)
	libc.musl-armv7.so.1 => /lib/ld-musl-armhf.so.1 (0x400000)
test: ELF 32-bit LSB pie executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-armhf.so.1, BuildID[sha1]=efdca87e0722dcb5ff5d34b78d71458be21f2f4a, with debug_info, not stripped
```

arm64
```sh
$ docker run -ti --rm --platform linux/arm64 alpine:3.20.3 sh -c 'apk add build-base && gcc -x c -Wl,--verbose=2 -v -static-pie <(echo "int main() {return 0;}") -o test && ldd test && file test'
...
attempt to open /usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../lib/libc.a succeeded
...
(/usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../lib/libc.a)__set_thread_area.lo
(/usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../lib/libc.a)default_attr.lo
/usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/libgcc.a
/usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/libgcc_eh.a
/usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../lib/libc.a
attempt to open /usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/crtendS.o succeeded
/usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/crtendS.o
attempt to open /usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../lib/crtn.o succeeded
/usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../lib/crtn.o
COLLECT_GCC_OPTIONS='-v' '-static-pie' '-o' 'test' '-march=armv8-a' '-mlittle-endian' '-mabi=lp64' '-dumpdir' 'test.'
	/lib/ld-musl-aarch64.so.1 (0xffffb8a10000)
test: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), static-pie linked, BuildID[sha1]=4ea26b9a349e649cde6ed6cd3ed149202aea1a05, with debug_info, not stripped
```

So looks like for some reason armv7 is looking for `.so`-files when using `-static-pic` 🤔 

Beware that `docker run --platform` seems to be a bit flaket, bug?, for me it sometimes runs a armv7 container even when using `--platform linux/arm64`. Maybe i'm doing something wrong?

**Comment by wader** [comment]
This might explain things https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106356 and it seems like at least that change is not in gcc master https://github.com/gcc-mirror/gcc/blob/master/gcc/config/arm/linux-elf.h#L69-L72

Injecting those link flags do seem to work:
```sh
$ docker run -ti --rm -v $PWD:$PWD -w $PWD --platform linux/arm/v7 alpine:3.20.1 sh -c 'apk add build-base && gcc -x c -static-pie -Wl,--verbose=2 -v -Wl,-Bstatic -Wl,-pie -Wl,--no-dynamic-linker -Wl,"-z text" <(echo "int main() {return 0;}") -o test && ldd test && file test'
attempt to open /usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../libc.a succeeded
....
/usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../libc.a
attempt to open /usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/crtendS.o succeeded
/usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/crtendS.o
attempt to open /usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../crtn.o succeeded
/usr/lib/gcc/armv7-alpine-linux-musleabihf/13.2.1/../../../crtn.o
COLLECT_GCC_OPTIONS='-static-pie' '-v' '-o' 'test'  '-mtune=generic-armv7-a' '-mfloat-abi=hard' '-mfpu=vfpv3-d16' '-mabi=aapcs-linux' '-mtls-dialect=gnu' '-mthumb' '-mlibarch=armv7-a+fp' '-march=armv7-a+fp' '-dumpdir' 'test.'
	/lib/ld-musl-armhf.so.1 (0x400000)
test: ELF 32-bit LSB pie executable, ARM, EABI5 version 1 (SYSV), static-pie linked, BuildID[sha1]=c006d2f3e3da10e6d9cf88a22adbc2116fa8f4cc, with debug_info, not stripped
```

**Comment by gtsteffaniak** [comment]
Yes I'm not an expert on this but isn't there a difference between static-pie and static?

If you don't specify static, "static-pie" attempts to make it "as static as it can", but might not

https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html

Not sure what the link you provided means, it's not enforcing static?

I need to dust of the old raspberry pi to test a native build.

**Comment by wader** [comment]
Both static and static PIE binaries are suppose to have no external dependencies. Static PIE is a quite new thing and means that the binary has "position independent code" so that some security features like ASLR can be used, i think you can more or less think about it as if the binary embeds the shared objects and it's own linker (ld). The security features is the reason i want to support it.

I guess solution if it's too messy to do at the moment might be to document that armv7 is just static and not static PIE and hopefully we can fix it in the future if/when gcc gets support.

**Comment by gtsteffaniak** [comment]
Yeah but maybe it's working as intended if armv7l doesn't support it (for whatever reason would be nice to figure out)

I'm referring to this line from the above GCC compiler link

```
-static-pie

Produce a static position independent executable on targets that support it.
```

Perhaps the compiler identified it's an incompatible environment, and defaults back to dynamically linked compilation?

**Comment by wader** [comment]
Yeah could be, but i can't see why any architecture that can support dynamic linking could not support static-pie. Based on the patch that as was submitted to the gcc mailing list and my testing i would guess there is just some glue missing between gcc and how it's calling the linker. So hopefully it will get fixed and maybe we could try to inject the same linker flags somehow?

**Comment by wader** [comment]
@mathieu-aubin hey! no worries! this Dockerfile should have always produced static-pie binaries with no external shared objects, the tests in the end of the build uses tools to check it and also the binary is test executed in a scratch container (empty image). But i know that some tools like `file` and `ldd` (gnu and musl has it's own implementations) can report quite confusing and weird things about them.

But i would probably try inject linker flags to produce static-pie on arm32 and if that is a mess to do we just produce normal static binaries for now and wait for proper support in gcc.

**Comment by gtsteffaniak** [comment]
I would also prefer a standard static-pie binary across all builds. I still need to confirm if a native armv7l raspberry pi has the same issue when compiling.

But I was discouraged at how loooong it takes to compile arm v7 , it could take several hours -- which would eat up the free tier of github actions daily limit.

**Comment by wader** [comment]
Yes it seems emulated armv7 even on a arm64 host it very slow :( A bit of a stretch but I do know that it is possible to affect qemu using the `QEMU_CPU` env variable. I've aware of one case were arm64 emulation was very slow due to how some versions of qemu emulates pointer auth and it was speed up by emulating a arm variant that don't support it.

**Comment by gtsteffaniak** [comment]
so, I tried building on a rasberry pi 32 bit os. Firstly, finding a 32 bit OS image that was actually 32 bit was so difficult. I flashed 4 different times 3 different OS's, including Raspbian official 32bit image, they all showed aarch64 in the `uname -a` 

Anyways, I went to the archive and downloaded an older image from 2020 and that one did indeed use a 32 bit kernel.

Now the Alpine Linux Docker build isn't working at the alpine steps, no matter what version of alpine I use or change the networking:

```
0.362 fetch https://dl-cdn.alpinelinux.org/alpine/v3.21/main/armv7/APKINDEX.tar.gz
0.407 WARNING: updating and opening https://dl-cdn.alpinelinux.org/alpine/v3.21/main: temporary error (try again later)
0.408 fetch https://dl-cdn.alpinelinux.org/alpine/v3.21/community/armv7/APKINDEX.tar.gz
0.454 WARNING: updating and opening https://dl-cdn.alpinelinux.org/alpine/v3.21/community: temporary error (try again later)
0.454 ERROR: Not continuing due to stale/unavailable repositories.Use --force-missing-repositories to continue.
```

I'm considering giving up on armv32 and just wont offer ffmpeg support for that variant anymore. What a shame. 

But I will still do a pull request that I think cleans up the ffmpeg configure command.

**Comment by wader** [comment]
> so, I tried building on a rasberry pi 32 bit os. Firstly, finding a 32 bit OS image that was actually 32 bit was so difficult. I flashed 4 different times 3 different OS's, including Raspbian official 32bit image, they all showed aarch64 in the `uname -a`
> 
> Anyways, I went to the archive and downloaded an older image from 2020 and that one did indeed use a 32 bit kernel.

This is a 64 bit rasberry pi that can run in 32 bit mode? 
 
> Now the Alpine Linux Docker build isn't working at the alpine steps, no matter what version of alpine I use or change the networking:
> 
> ```
> 0.401 fetch https://dl-cdn.alpinelinux.org/alpine/v3.21/main/armv7/APKINDEX.tar.gz
> 0.502 WARNING: fetching https://dl-cdn.alpinelinux.org/alpine/v3.21/main: temporary error (try again later)
> 0.502 fetch https://dl-cdn.alpinelinux.org/alpine/v3.21/community/armv7/APKINDEX.tar.gz
> 0.547 WARNING: fetching https://dl-cdn.alpinelinux.org/alpine/v3.21/community: temporary error (try again later)
> ```

Huh strange, is that any networking or just apt?

> I'm considering giving up on armv32 and just wont offer ffmpeg support for that variant anymore. What a shame.

Yeah it's a bit of sad situation, but thanks for trying and we at least dugg up with new things that might be useful if/when some way of building it shows up. I'm hoping someone will eventually provide armv7 action runners.

> But I will still do a pull request that I think cleans up the ffmpeg configure command.

👍
